### PR TITLE
Add format for "placeholder" items

### PR DESCRIPTION
**Pull request by jamiecobbett**
We have around ~~149,161~~ 153,236 placeholder items, but we don't have a schema for the
format. I've developed this against the placeholders produced by (mainstream)
publisher and whitehall. Some small changes will be required for publisher to
supply additional fields like "locale", but Whitehall's placeholders are already valid against this schema.

Once this is merged, I'll open PRs to test Whitehall's placeholders and to add schema tests to Publisher.

### CODE REVIEW

File: formats/placeholder/publisher/schema.json [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+      }
+    },
+    "guid_list": {
+      "type": "array",
+      "items": {
+        "pattern": "^[a-f0-9]{8}-[a-f0-9]{4}-[1-5][a-f0-9]{3}-[89ab][a-f0-9]{3}-[a-f0-9]{12}$"
+      }
+    }
+  }
+}
```
Comment (rboulton) [inline formats/placeholder/publisher/schema.json]: Shouldn't this have a newline at the end of the file?

Comment (jamiecobbett) [inline formats/placeholder/publisher/schema.json]: These files are generated by the `make` task. It should be possible to change the code which generates them to add a trailing newline, but that's outside of the scope.

Comment (rboulton) [inline formats/placeholder/publisher/schema.json]: Ah, no problem, then.

File: formats/placeholder/frontend/schema.json [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    "title": {
+      "type": "string"
+    },
+    "description": {
+      "type": "string"
+    },
+    "format": {
+      "type": "string",
+      "enum": [
+        "placeholder"
```
Comment (tekin) [inline formats/placeholder/frontend/schema.json]: This will mean that any placeholder items with the prefix format (e.g. "placeholder_organisation") will not pass validation. Can this be made a pattern match instead?

File: formats/placeholder/publisher/details.json [hunk 3]
```diff
@@ -0,0 +1,39 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "type": "object",
+  "additionalProperties": false,
```
Comment (alext) [inline formats/placeholder/publisher/details.json]: Does this mean that placeholder items are restricted in what can be in the details hash?  I'm not sure that's necessarily what we want.

A placeholder item could readily have the full details that the eventual non-placeholder format will have, and only be using the placeholder variant to prevent routes being registered during the transition phase.

File: formats/placeholder/frontend/schema.json (hunk 2 above)
Comment (jamiecobbett) [inline formats/placeholder/frontend/schema.json]: Good idea.

Comment (jamiecobbett) [inline formats/placeholder/frontend/schema.json]: Unfortunately, that would require a code change as the `format` enum is [automatically added based upon the format](https://github.com/alphagov/govuk-content-schemas/blob/master/lib/govuk_content_schemas/schema_combiner.rb#L23). Given that you're creating a precedent for variable values for `format`, would you be happy to try it in a separate PR?

Comment (tekin) [inline formats/placeholder/frontend/schema.json]: This isn't a new precedence, there is already be a bunch of content items with the "placeholder_FORMAT"-style of format. For example, see https://github.com/alphagov/whitehall/blob/master/app/presenters/publishing_api_presenters/placeholder.rb

File: formats/placeholder/publisher/details.json (hunk 3 above)
Comment (jamiecobbett) [inline formats/placeholder/publisher/details.json]: Generally, we disallow undefined keys for good usability reasons and it would be a shame to lose that. For Whitehall case studies, we are doing something very similar by sending the item with a format of "case_study", but a rendering_app of "whitehall-frontend", rather than "government-frontend". Maybe we can defer what you suggest until we need it?

File: formats/placeholder/frontend/schema.json (hunk 2 above)
Comment (jamiecobbett) [inline formats/placeholder/frontend/schema.json]: Ah, you're right, sorry. I need to add another schema validation to my Whitehall branch.

Comment (jamiecobbett) [inline formats/placeholder/frontend/schema.json]: OK, finally got it working in 400b4c5.

**Comment by jamiecobbett** [comment]
Any more for any more?

File: formats/placeholder/frontend/schema.json [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    "title": {
+      "type": "string"
+    },
+    "description": {
+      "type": "string"
+    },
+    "format": {
+      "type": "string",
+      "pattern": "^(placeholder|placeholder_.+)$",
+      "description": "Should be of the form 'placeholder_my_format_name'. 'placeholder' is allowed for backwards compatibility."
```
Comment (tekin) [inline formats/placeholder/frontend/schema.json]: I may be wrong, but 'placeholder' isn't just there for backwards compatibility AFAIK. It's just as valid as 'placeholder_FORMAT'. The main difference is that the latter allows the content-register to identify the content and add it to the register. I can see situations where we may want to push placeholder content out without this happening. @alext?

File: formats/placeholder/frontend/schema.json [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    },
+    "title": {
+      "type": "string"
+    },
+    "description": {
+      "type": "string"
+    },
+    "format": {
+      "type": "string",
+      "pattern": "^(placeholder|placeholder_.+)$",
```
Comment (tekin) [inline formats/placeholder/frontend/schema.json]: :+1:

**Comment by tekin** [comment]
Minor comment on the description for the format, but otherwise looks good to me :+1:.
